cpu/atxmega/periph/timer.c

```diff
@@ -0,0 +1,226 @@
+ *
+ * @author      Josua Arndt <jarndt@ias.rwth-aachen.de>
+ * @author      Gerson Fernando Budke <nandojve@gmail.com>
+ *
+ * @}
+ */
+
+#include <avr/interrupt.h>
+
+#include "cpu.h"
+#include "thread.h"
+
+#include "periph/timer.h"
+
+#include "board.h"
+#include "periph_conf.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/**
+ * @brief   All timers have three channels
+ */
+#define CHANNELS            (2)
+
+/**
+ * @brief   We have 5 possible prescaler values
+ */
+#define PRESCALE_NUMOF      (7U)
+
+/**
+ * @brief   Possible prescaler values, encoded as 2 ^ val
+ */
+static const uint8_t prescalers[] = { 0, 1, 2, 3, 6, 8, 10 };
+
+/**
+ * @brief   Timer state context
+ */
+typedef struct {
+    TC1_t *dev;             /**< timer device */
+    timer_cb_t cb;          /**< interrupt callback */
+    void *arg;              /**< interrupt callback argument */
+    uint8_t mode;           /**< remember the configured mode */
+    uint8_t isrs;           /**< remember the interrupt state */
+    int_lvl_t int_lvl;      /**< interrupt level */
+} ctx_t;
+
+/**
+ * @brief   Allocate memory for saving the device states
+ * @{
+ */
+#ifdef TIMER_NUMOF
+static ctx_t ctx[] = {
+#ifdef TIMER_0
+    { TIMER_0, NULL, NULL, 0, 0, TIMER_0_INT_LVL },
+#endif
+#ifdef TIMER_1
+    { TIMER_1, NULL, NULL, 0, 0, TIMER_1_INT_LVL },
+#endif
+#ifdef TIMER_2
+    { TIMER_2, NULL, NULL, 0, 0, TIMER_2_INT_LVL },
+#endif
+#ifdef TIMER_3
+    { TIMER_3, NULL, NULL, 0, 0, TIMER_3_INT_LVL },
+#endif
+};
+#else
+/* fallback if no timer is configured */
+static ctx_t *ctx[] = { { NULL } };
+#endif
+/** @} */
+
+/**
+ * @brief Setup the given timer
+ */
+int timer_init(tim_t tim, unsigned long freq, timer_cb_t cb, void *arg)
+{
+    DEBUG("timer.c: freq = %ld, Core Clock = %ld\n", freq, CLOCK_CORECLOCK);
+
+    uint8_t pre = 0;
+
+    /* make sure given device is valid */
+    if (tim >= TIMER_NUMOF) {
+        return -1;
+    }
+
+    /* figure out if freq is applicable */
+    for (; pre < PRESCALE_NUMOF; pre++) {
+        if ((CLOCK_CORECLOCK >> prescalers[pre]) == freq) {
+            break;
+        }
+    }
+    if (pre == PRESCALE_NUMOF) {
+        DEBUG("timer.c: prescaling failed!\n");
+        return -1;
+    }
+
+    /* stop and reset timer */
+    ctx[tim].dev->CTRLA = 0;                    /* Stop */
+    ctx[tim].dev->CTRLFSET = TC_CMD_RESET_gc;   /* Force Reset */
+
+    /* save interrupt context and timer Prescaler */
+    ctx[tim].cb = cb;
+    ctx[tim].arg = arg;
+    ctx[tim].mode = (0x0F & (pre + 1));
+
+    /* enable timer with calculated prescaler */
+    ctx[tim].dev->CTRLA = ctx[tim].mode;
+    DEBUG("timer.c: prescaler set to %d \n", ctx[tim].dev->CTRLA );
+
+    return 0;
+}
+
+int timer_set_absolute(tim_t tim, int channel, unsigned int value)
+{
+    if (channel >= CHANNELS) {
+        return -1;
+    }
+    DEBUG("timer_set_absolute channel %d to %u\n", channel, value );
+
+    // Compare or Capture Disable
```

Comment:
```suggestion
    /* Compare or Capture Disable */
```

let's keep the comment style consistent :)